034-Digit_factorials.py

```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf-8 -*-
+'''
+145 is a curious number, as 1! + 4! + 5! = 1 + 24 + 120 = 145.
+
+Find the sum of all numbers which are equal to the sum of the factorial of their digits.
+
+Note: as 1! = 1 and 2! = 2 are not sums they are not included.
+'''
+import time
+
+def factorial(n):
+    # factorial of zero is 1
+    if n == 0:
+        return 1
+    else:
+        return n * factorial(n - 1)
+
+# 9! = 362880 has six digits. So if a number has '9' inside it, in order to have
+# sum of factorials equals the number, the max limit need to be a 6 digit number.
+# 6 * 9! = 2177280 has 7 digits. So the upper limit should be 7 * 9! = 2540160.
+total = 0
+start = time.time()
+
+# cache the factorials
+factorials = [factorial(i) for i in range(0, 10)]
```

Comment:
Your approach still has O(n2) time and O(n2) space complexity.
It can be further optimized:
factorials = [1]
for i in range (1, 10):
  factorials.append(factorials[-1] * i)
This is only O(n) time and O(n) space complexity.